modules/videoio/src/cap_ffmpeg_hw.hpp

```diff
@@ -0,0 +1,297 @@
+    if (hw_type == AV_HWDEVICE_TYPE_QSV) {
+#ifdef _WIN32
+        child_type = AV_HWDEVICE_TYPE_DXVA2;
+#else
+        child_type = AV_HWDEVICE_TYPE_VAAPI;
+#endif
+    }
+
+    AVBufferRef* hw_device_ctx = NULL;
+    char device[128] = "";
+    char* pdevice = NULL;
+    if (hw_device >= 0 && hw_device < 100000) {
+        if (child_type == AV_HWDEVICE_TYPE_VAAPI) {
+            snprintf(device, sizeof(device), "/dev/dri/renderD%d", 128 + hw_device);
+        } else {
+            snprintf(device, sizeof(device), "%d", hw_device);
+        }
+        pdevice = device;
+    }
+    av_hwdevice_ctx_create(&hw_device_ctx, child_type, pdevice, NULL, 0);
+    if (hw_device_ctx) {
+        if (!hw_check_device(hw_device_ctx, hw_type)) {
+            av_buffer_unref(&hw_device_ctx);
+            return NULL;
+        }
+        if (hw_type != child_type) {
+            AVBufferRef *derived_ctx = NULL;
+            av_hwdevice_ctx_create_derived(&derived_ctx, hw_type, hw_device_ctx, 0);
+            av_buffer_unref(&hw_device_ctx);
+            return derived_ctx;
+        } else {
+            return hw_device_ctx;
+        }
+    } else {
+        return NULL;
+    }
+}
+
+AVBufferRef* hw_create_frames(struct AVCodecContext* ctx, AVBufferRef *hw_device_ctx, int width, int height, AVPixelFormat hw_format)
+{
+    AVBufferRef *hw_frames_ref = nullptr;
+    if (ctx) {
+        avcodec_get_hw_frames_parameters(ctx, hw_device_ctx, hw_format, &hw_frames_ref);
+    }
+    if (!hw_frames_ref) {
+        hw_frames_ref = av_hwframe_ctx_alloc(hw_device_ctx);
+    }
+    if (!hw_frames_ref) {
+        CV_LOG_DEBUG(NULL, "Failed to create HW frame context");
+        return NULL;
+    }
+    AVHWFramesContext *frames_ctx = (AVHWFramesContext *)(hw_frames_ref->data);
+    frames_ctx->width = width;
+    frames_ctx->height = height;
+    if (frames_ctx->format == AV_PIX_FMT_NONE)
+        frames_ctx->format = hw_format;
+    if (frames_ctx->sw_format == AV_PIX_FMT_NONE)
+        frames_ctx->sw_format = HW_DEFAULT_SW_FORMAT;
+    if (frames_ctx->initial_pool_size == 0)
+        frames_ctx->initial_pool_size = HW_DEFAULT_POOL_SIZE;
+    if (av_hwframe_ctx_init(hw_frames_ref) < 0) {
+        CV_LOG_DEBUG(NULL, "Failed to initialize HW frame context");
+        av_buffer_unref(&hw_frames_ref);
+        return NULL;
+    }
+    return hw_frames_ref;
+}
+
+static bool hw_check_codec(AVCodec* codec, AVHWDeviceType hw_type) {
+    // disable MJPG HW encoder due to low encoding quality
+    if (AV_CODEC_ID_MJPEG == codec->id && av_codec_is_encoder(codec) && AV_HWDEVICE_TYPE_NONE != hw_type)
+        return false;
+    // disable AV1 HW decoder due to SW fallback issue
+    if (AV_CODEC_ID_AV1 == codec->id && av_codec_is_decoder(codec) && AV_HWDEVICE_TYPE_NONE != hw_type)
+        return false;
+    // disable VP8 VAAPI decoder/encoder
+    if (AV_CODEC_ID_VP8 == codec->id && AV_HWDEVICE_TYPE_VAAPI == hw_type)
+        return false;
+    return true;
```

Comment:
I extended usage of environment variable OPENCV_FFMPEG_CAPTURE_OPTIONS and introduced OPENCV_FFMPEG_WRITER_OPTIONS.
Besides disabling specific codecs, it could be used for disabling specific acceleration type
```
export OPENCV_FFMPEG_CAPTURE_OPTIONS="disable_codecs;.vaapi"
export OPENCV_FFMPEG_WRITER_OPTIONS="disable_codecs;.vaapi"
```
or all HW codecs
```
export OPENCV_FFMPEG_CAPTURE_OPTIONS="disable_codecs;hw"
export OPENCV_FFMPEG_WRITER_OPTIONS="disable_codecs;hw"
```
or not disable (none as value not matching any codec)
```
export OPENCV_FFMPEG_CAPTURE_OPTIONS="disable_codecs;none"
export OPENCV_FFMPEG_WRITER_OPTIONS="disable_codecs;none"
```
Syntax schema inherited from OPENCV_FFMPEG_CAPTURE_OPTIONS with separation symbols ; and |, looks little weird:)
```
export OPENCV_FFMPEG_CAPTURE_OPTIONS="rtsp_transport;udp|disable_codecs;av1.vaapi,av1_qsv"
```

Reply:
Great! Thank you

> or not disable at all (none as value not matching any codec)

Awesome :)
This is extremely useful for testing.


>  with separation symbols ; and |, looks little weird:)

Agreed... but it was done 3 years ago (#9292), so we need to keep compatibility here.
Hopefully we will add "string" properties/parameters soon through normal API with improved separation symbols.

FFmpeg internally uses different sets internally, so there is no gold approach:
- "=", ":" (8 times)
- "=", "," (2 times)
- some special like "@", "#", AV_DICT_MULTIKEY